src/names/name-resolution.md

```diff
@@ -1,4 +1,416 @@
+r[names.resolution]
 # Name resolution
 
+r[names.resolution.intro]
+
+_Name resolution_ is the process of tying paths and other identifiers to the
+declarations of those entities. Names are segregated into different
+[namespaces], allowing entities in different namespaces to share the same name
+without conflict. Each name is valid within a [scope], or a region of source
+text where that name may be referenced. Access to certain names may be
+restricted based on their [visibility].
+
+Name resolution is split into three stages throughout the compilation process.
+The first stage, Expansion-time resolution, resolves all [use declarations] and
+[macro invocations]. The second stage, Primary resolution, resolves all names
+that have not yet been resolved that do not depend on type information to
+resolve. The last stage, Type-relative resolution, resolves the remaining names
+once type information is available.
+
+> Note
+>
+> * Expansion-time resolution is also known as "Early Resolution"
+> * Primary resolution is also known as "Late Resolution"
+
+r[names.resolution.expansion]
+## Expansion-time name resolution
+
+r[names.resolution.expansion.intro]
+
+Expansion-time name resolution is the stage of name resolution necessary to
+complete macro expansion and fully generate a crate's AST. This stage requires
+the resolution of macro invocations and use declarations. Resolving use
+declarations is required to resolve [path-based scope] macro invocations.
+Resolving macro invocations is required in order to expand them.
+
+The expansion process is iterative, alternately resolving imports, resolving
+and expanding macro invocations, then repeating until there are no further
+macros invocations to resolve. Once this process is completed all the imports
+are resolved again to ensure that the macro expansion process did not introduce
+any new ambiguious imports.
```

Comment:
do we want to talk about this? feels like an implementation detail but
also really helps to understand certain kinds of ambiguity errors that users
can run into.